Create an Event loop runner

Jamie-Chang opened the pull request:
closes #7

Jamie-Chang committed:
Create an Event loop runner

martindurant commented:
What refreshingly modern-looking python code!

I have yet to try this in earnest, but here are a couple of thoughts:
- using the async context makes sense compared to the original concept of aiointerpreters, but I think fsspec would want something that can hang around idle; so it would be better to separate out start and stop methods.
- call_soon_threadsafe and queue operations are probably relatively expensive because they will involve locks/synchronisation. If it could be arranged to pass lists of coroutines/results around, I would expect that to be much more performant.

martindurant commented:
Additionally: the code assumes py3.14, which is when the module concurrent.interpreters became available. However, the underlying subinterpreters have been around since (I think) 3.12, so it may be worth considering whether we can backport, when we already have something worth sharing.

Jamie-Chang commented:
> Additionally: the code assumes py3.14, which is when the module concurrent.interpreters became available. However, the underlying subinterpreters have been around since (I think) 3.12, so it may be worth considering whether we can backport, when we already have something worth sharing.

3.12/3.13 compatibility is possible but we'll need to do some research, AFAIK, some of the implementation details of interpreters were changed leading up to the 3.14 release

Jamie-Chang commented:
> * using the async context makes sense compared to the original concept of aiointerpreters, but I think fsspec would want something that can hang around idle; so it would be better to separate out start and stop methods.

@martindurant this is doable, but I think I would need a clearer idea of the lifetime of the "interpreterpool", I suppose we would attach the interpreter pool to the FileSystem class itself and keep it active for the lifetime of the class

Jamie-Chang commented:
Another open issue is that my mechanism for passing coroutine objects is very limited, it only handles top level coroutine functions and not methods. We would probably need to pickle the FileSystem instance for the fsspec usecase, and this would have to be done for each batch.

martindurant commented:
See https://github.com/Jamie-Chang/aiointerpreters/pull/8

martindurant commented:
> I think I would need a clearer idea of the lifetime of the "interpreterpool"

In fsspec's current async implementation, the loop runs essentially for the whole session and is shared by all async filesystem instances. That can sometimes lead to warnings at shurdown time (loop closes and thread ends before objects are cleaned up) but we have dealt with 99% of them I think. There's no reason an interpreter async pool shouldn't live the whole session, it shouldn't hold a lot of resources when idle.

> We would probably need to pickle the FileSystem instance for the fsspec usecase

All fsspec instances are pickleable, and also the instances are cached, so that if you attempt to instantiate a set of class/init args already seen, there is no initialisation cost. If we send batches of coroutine methods of one instance, then pickle is smart enough to only serialise the instance once.

Jamie-Chang commented:
Cool I think I can make it work then

Jamie-Chang committed:
Make event loop runner more flexible

Jamie-Chang commented:
@martindurant update:
- pickling is more robust, support is added for method derived coroutines that would fit the usecase of fsspec 
- added start and stop methods
- added create_batch method 

3.12/3.13 backport will be investigated separately 

let me know if there's any other concerns with this
